File: algorithm/src/leetcode/ruby/ZigzagConversion_6.rb
# 6 ruby convert

# @param {String} s
# @param {Integer} num_rows
# @return {String}
def convert(s, num_rows)
  return s if num_rows == 1 || s.length == 0 || s.length == 1

  block = {}
  (0...num_rows).each do |idx|
    block[idx] = ""
  end

  isRight = true
  idx, row = 0, 0
  while(idx < s.length)
    word = s[idx]

    block[row] += word
    if isRight
      row += 1
      if(row == num_rows-1)
        isRight = false
      end
    else
      row -= 1
      if(row == 0)
        isRight = true
      end
    end
    idx += 1
  end

  answer = ""
  block.values.flatten.map { |str| answer += str }.last
end

s = "ABC"
num_rows = 1

p convert(s, num_rows)

File: java/problem_solving/src/leetcode/ZigZagConversion_6.java
package leetcode;

// #6 - ZigZag Conversion
import java.util.ArrayList;
import java.util.List;

public class ZigZagConversion_6 {
	public static void main(String[] args) {
		String s = "PAYPALISHIRING";
		int numRows = 3;
		System.out.println("convert(s, numRows) = " + convert(s, numRows));
	}

	public static String convert(String s, int numRows) {
		int len = s.length();
		if(numRows == 1 || len == 0 || len == 1) return s;

		List<StringBuilder> builders = new ArrayList<>();
		for (int i = 0; i < numRows; i++) {
			builders.add(i, new StringBuilder());
		}

		boolean isRight = true;
		int idx = 0;
		int row = 0;
		while (idx < s.length()) {
			char word = s.charAt(idx);
			StringBuilder rowBuilder = builders.get(row);
			rowBuilder.append(word);
			if (isRight) {
				row++;
				if (row == numRows - 1) {
					isRight = false;
				}
			} else {
				row--;
				if (row == 0) {
					isRight = true;
				}
			}
			idx++;
		}

		StringBuilder answer = new StringBuilder();
		for (int i = 0; i < builders.size(); i++) {
			answer.append(builders.get(i).toString());
		}
		return answer.toString();
	}
}
